Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_wfssrv.cpp

```diff
@@ -0,0 +1,243 @@
+    return_error_code = WFS_EEMPTY;
+    break;
+
+  case IOCTL_WFS_GLOB_END:
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_GLOB_END(%d)", command);
+    Memory::Memset(buffer_out, 0, buffer_out_size);
+    break;
+
+  case IOCTL_WFS_OPEN:
+  {
+    u32 mode = Memory::Read_U32(buffer_in);
+    u16 path_len = Memory::Read_U16(buffer_in + 0x20);
+    std::string path = Memory::GetString(buffer_in + 0x22, path_len);
+
+    u16 fd = GetNewFileDescriptor();
+    FileDescriptor* fd_obj = &m_fds[fd];
+    fd_obj->in_use = true;
+    fd_obj->path = path;
+    fd_obj->mode = mode;
+    fd_obj->position = 0;
+
+    if (!fd_obj->Open())
+    {
+      ERROR_LOG(WII_IPC_HLE, "IOCTL_WFS_OPEN(%s, %d): error opening file", path.c_str(), mode);
+      ReleaseFileDescriptor(fd);
+      return_error_code = -1;  // TODO(wfs): proper error code.
+      break;
+    }
+
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_OPEN(%s, %d) -> %d", path.c_str(), mode, fd);
+    Memory::Write_U16(fd, buffer_out + 0x14);
+    break;
+  }
+
+  case IOCTL_WFS_CLOSE:
+  {
+    u16 fd = Memory::Read_U16(buffer_in + 0x4);
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_CLOSE(%d)", fd);
+    ReleaseFileDescriptor(fd);
+    break;
+  }
+
+  case IOCTL_WFS_READ:
+  {
+    u32 addr = Memory::Read_U32(buffer_in);
+    u16 fd = Memory::Read_U16(buffer_in + 0xC);
+    u32 size = Memory::Read_U32(buffer_in + 8);
+
+    FileDescriptor* fd_obj = FindFileDescriptor(fd);
+    if (fd_obj == nullptr)
+    {
+      ERROR_LOG(WII_IPC_HLE, "IOCTL_WFS_READ: invalid file descriptor %d", fd);
+      return_error_code = -1;  // TODO(wfs): proper error code.
+      break;
+    }
+
+    size_t read_bytes;
+    if (!fd_obj->file.ReadArray(Memory::GetPointer(addr), size, &read_bytes))
+    {
+      return_error_code = -1;  // TODO(wfs): proper error code.
+      break;
+    }
+    fd_obj->position += read_bytes;
+
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_READ: read %zd bytes from FD %d (%s)", read_bytes, fd,
+             fd_obj->path.c_str());
+    return_error_code = static_cast<int>(read_bytes);
+    break;
+  }
+
+  default:
+    // TODO(wfs): Should be returning -3. However until we have everything
+    // properly stubbed it's easier to simulate the methods succeeding.
+    WARN_LOG(WII_IPC_HLE, "%s unimplemented IOCtl(0x%08x, size_in=%08x, size_out=%08x)\n%s\n%s",
+             m_name.c_str(), command, buffer_in_size, buffer_out_size,
+             HexDump(Memory::GetPointer(buffer_in), buffer_in_size).c_str(),
+             HexDump(Memory::GetPointer(buffer_out), buffer_out_size).c_str());
+    Memory::Memset(buffer_out, 0, buffer_out_size);
+    break;
+  }
+
+  Memory::Write_U32(return_error_code, command_address + 4);
+  return GetDefaultReply();
+}
+
+IPCCommandResult CWII_IPC_HLE_Device_usb_wfssrv::IOCtlV(u32 command_address)
+{
+  SIOCtlVBuffer command_buffer(command_address);
+  ERROR_LOG(WII_IPC_HLE, "IOCtlV on /dev/usb/wfssrv -- unsupported");
+  Memory::Write_U32(FS_EINVAL, command_address + 4);
```

Comment:
This should be IPC_EINVAL. I'm sorry, but when I cleaned up the return code constants, I renamed FS_EINVAL to IPC_EINVAL (for -4) and FS_EINVAL is now the real EINVAL code for filesystem code.